ui/hooks/gator-permissions/useGatorPermissionTokenInfo.ts

```diff
@@ -0,0 +1,404 @@
+// Exported for testing purposes
+export const tokenInfoPromiseCache = new Map<string, Promise<TokenInfo>>();
+
+/**
+ * Fetch token info from API with fallback to on-chain data.
+ *
+ * @param address - Token address
+ * @param chainId - Chain ID in hex format
+ * @param allowExternalServices - Whether to use external API services
+ * @returns Token information with source indicator
+ */
+async function fetchTokenInfoWithFallback(
+  address: string,
+  chainId: Hex,
+  allowExternalServices: boolean,
+): Promise<{ tokenInfo: TokenInfo; source: 'api' | 'onchain' }> {
+  const cacheKey = `${chainId}:${address.toLowerCase()}`;
+
+  // Check if there's already a request in flight and await it
+  const existingPromise = tokenInfoPromiseCache.get(cacheKey);
+  if (existingPromise) {
+    const tokenInfo = await existingPromise;
+    // Determine source from cached data (if it has image, likely from API)
+    return {
+      tokenInfo,
+      source: tokenInfo.image ? 'api' : 'onchain',
+    };
+  }
+
+  // Create a promise for this fetch operation
+  const fetchPromise = (async () => {
+    let symbol: string | undefined;
+    let decimals: number | undefined;
+    let name: string | undefined;
+    let image: string | undefined;
+
+    // Tier 1: Try API if external services are allowed
+    if (allowExternalServices) {
+      try {
+        const metadata = await fetchAssetMetadata(address, chainId);
+        if (metadata) {
+          symbol = metadata.symbol;
+          decimals = metadata.decimals;
+          // Note: fetchAssetMetadata returns 'address' and 'image', not 'name'
+          image = metadata.image;
+        }
+      } catch (error) {
+        log.warn('Failed to fetch token metadata from API', {
+          address,
+          chainId,
+          error: error instanceof Error ? error.message : String(error),
+        });
+      }
+    }
+
+    // Tier 2: Fall back to on-chain data if needed
+    if (!symbol || decimals === null || decimals === undefined) {
+      try {
+        const details = await getTokenStandardAndDetailsByChain(
+          address,
+          undefined,
+          undefined,
+          chainId,
+        );
+
+        const decRaw = details?.decimals as string | number | undefined;
+        if (typeof decRaw === 'number') {
+          decimals = decRaw;
+        } else if (typeof decRaw === 'string') {
+          // Handle hex strings (with 0x prefix) or decimal strings
+          const trimmed = decRaw.trim();
+          let parsed: number;
+
+          if (trimmed.startsWith('0x') || trimmed.startsWith('0X')) {
+            // Parse as hex (strip the 0x prefix)
+            parsed = parseInt(trimmed.slice(2), 16);
+          } else {
+            // Parse as decimal
+            parsed = parseInt(trimmed, 10);
+          }
+
+          // Only accept valid, finite, non-negative integers
+          if (Number.isFinite(parsed) && parsed >= 0) {
+            decimals = parsed;
+          }
```

Comment:
there's a handy dandy `hexToNumber` exported from @metamask/utils that would replace a bunch of this logic